numba/core/types/npytypes.py

```diff
@@ -587,3 +587,15 @@ def strides(self):
     @property
     def key(self):
         return self.dtype, self.shape
+
+    def copy(self, dtype=None, ndim=None, layout=None, readonly=None):
+        if dtype is None:
+            dtype = self.dtype
+        if ndim is None:
+            ndim = self.ndim
+        if layout is None:
+            layout = self.layout
+        if readonly is None:
+            readonly = not self.mutable
+        cls = type(self)
+        return cls(dtype=dtype, shape=self.shape)
```

Comment:
Should `ndim`, `layout`, and `readonly` be used here somehow?

Reply:
Or, is this change absolutely necessary? I can't quite tell why this omission doesn't make tests fail, or maybe it isn't hit by any of the tests?

Reply:
After some more investigation:

- I think this is copy-pasted from `Array` above, which is why it is the way it presently is?
- It seems this function is only called here: https://github.com/numba/numba/blob/master/numba/core/typing/arraydecl.py#L153

I think it would be more natural to accept the parameters passed by `get_array_index_type()` and validate that they match the existing shape of the `NestedArray`, since we can't reasonably change it for a copy. This is implemented in https://github.com/gmarkall/numba/commit/7d21f75fbb2807860e4a279995e29c4c6003bd3a in the branch https://github.com/gmarkall/numba/tree/feature/nested_arrays_getitem - what do you think about this approach? Would you be happy with adding it to this PR?

It feels like the right thing to do would be

Reply:
you make a great point. I looked into this and thought of a new test, which I added in this commit: https://github.com/numba/numba/pull/7359/commits/ac492bffd34bc9c2f1152aeab62a960f204d695f

Interestingly, both the first version and the one in your branch fail the test. However, if I remove them (and let the base class method take care of this) the test passes. I need to think what this means. Maybe this operation should be handled differently, or not allowed, but I'm not sure. I'm trying to think of potential negative side effects of allowing the base class method take care of this case.

Reply:
An `Array` is more general than a `NestedArray`, so in terms of typing my gut feel is that this is safe. In terms of lowering, I'm wondering whether this would lead to any surprises with NRT, so I'd need to think a little more about that.